Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.hpp

```diff
@@ -0,0 +1,68 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// Acts include(s).
+#include "Acts/Seeding/InternalSeed.hpp"
+#include "Acts/Seeding/InternalSpacePoint.hpp"
+#include "Acts/Seeding/SeedFilter.hpp"
+#include "Acts/Seeding/SeedfinderConfig.hpp"
+
+// Sycl plugin include(s).
+#include "Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp"
+#include "Acts/Plugins/Sycl/Seeding/detail/Types.h"
+#include "Acts/Plugins/Sycl/Utilities/QueueWrapper.hpp"
+
+namespace Acts::Sycl {
+
+template <typename external_spacepoint_t>
+class Seedfinder {
+ public:
+  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
+             const Acts::Sycl::DeviceExperimentCuts& cuts);
+
+  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
+             const Acts::Sycl::DeviceExperimentCuts& cuts,
+             Acts::Sycl::QueueWrapper&& wrappedQueue);
+
+  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
+             const Acts::Sycl::DeviceExperimentCuts& cuts,
+             const Acts::Sycl::QueueWrapper& wrappedQueue);
```

Comment:
I see your intent here, but I'm not convinced at all that this would result in more readable code than what I had in mind...

Sure, this code can work, but to me it just makes ownership of a queue a lot harder to follow than what I was proposing. And I also didn't see many cases of people using the `&&` operator in anything else than the move constructor or move assignment operator of a class.

Normally you would have just a single constructor to this class.

```c++
  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
             const Acts::Sycl::DeviceExperimentCuts& cuts,
             Acts::Sycl::QueueWrapper wrappedQueue);
```

With this, the user could either pass an object to the constructor "by value" (in which case `QueueWrapper`'s copy constructor would be used), or using `std::move(...)` (in which case `QueueWrapper`'s move constructor would be used).

But it still doesn't solve my original issue. :frowning: What if the client code already has a `cl::sysl::queue` object that they want the seedfinder to use? How would they pass **that** in?

Reply:
I didn't know that, but I'll change the constructors. :)

I was thinking a lot about the original issue. I think it is better to manage (creating and destroying) queue-s through a class.

Our intention with hiding sycl headers in the first place, was that we can hide it's implementation details. With expecting a pointer to the queue, we expect that the other TU see's sycl headers (they would definitely need them for destroying the queue object).

(Also this way, we don't have to forward declare the type every time we use it.)

I know it's not clear where the ownership is in this version, but I think it's logical. If you create a new object, you're the owner. If you move, you're also owner. If you make a copy, than you don't own anything.
This also allows dereferencing null pointer, in case you make a copy and the copy doesn't know about the pointer's death. Still, the other version you proposed would have this issue as well.

Reply:
> I would also suggest that you move your example code from `Tests/UnitTests/Plugins/Sycl` to `Tests/IntegrationTests/Sycl`. Unit tests should be automated tests using the chosen unit test framework, which your example is not. Integration tests are not as tightly defined and better match your use case.

I think this is a good idea. Let me just point out that @czangela didn't come up with this location by herself. She was copying what I did, and what @beomki-yeo did before me. :stuck_out_tongue:
  - https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Plugins/Cuda/Seeding
  - https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Plugins/Cuda/Seeding2

And we just copied what you did for the CPU implementation. :stuck_out_tongue:

https://github.com/acts-project/acts/tree/master/Tests/UnitTests/Core/Seeding

So yes, if we want to move the executable to the integration tests, let's do that. We will just need to remember to move all the other executables later on as well.

Reply:
I have no idea why GitHub put this message here. :confused: